### DISCUSSION TITLE
Should I divide sub-classes?

### DISCUSSION THREAD

**Discussion by fluffytid** [ultralytics/ultralytics#10399] [Q&A]
Hi, I am working with playing cards. And this weeked I encountered a problem: The game I recorded to run on my model was played with a different style of playing cards, with the most notable difference being the 7s:

![image](https://github.com/ultralytics/ultralytics/assets/35581011/6df79884-6054-4d5c-81bb-ad59cc0f5a17)
![image](https://github.com/ultralytics/ultralytics/assets/35581011/2c6cf032-352e-4138-a3e1-0a5475004b89)

When I run my model on the first image, the card was considered background, not even a bad read from another class.

There are also some minor differences on other cards, specially figures that produced some errors, but I believe I can solve them with more samples.

Note that I know for a fact that my model has focused on the centre icons of the card and not the number 7 on the corners which is ignored for the most part. The way it tells a card from another is by the arrangement of the icons.

My question is: Should I add the new images to the same model, or should I differentiate with a new model for this type of deck. Or perhaps I should differentiate creating a 53rd class so that the different 7s don't make a fight.

**Comment by pderrenger** [comment]
@fluffytid hi! 🌟 The challenge with different styles of playing cards is indeed intriguing. Given your model focuses on the center icons rather than the corner numbers, and considering the notable difference in the style of 7s, adding more samples of the new style to your existing classes might cause confusion unless the model can generalize well across both styles. 

However, before splitting them into separate models or creating a new class for every different style, which could potentially inflate the number of classes unnecessarily, I recommend first trying to augment your dataset with more samples of the new style. This includes the different 7s and other minor variations you've mentioned. Data augmentation helps the model generalize better across variations within the same class.

If after augmenting your dataset you still encounter significant misclassifications, you might consider adding a new class for the distinctly different style of 7s as a 53rd class. This way, the model can learn the specific features of these outliers without disrupting the existing class structure.

Here's a snippet on how you might add the new images to your dataset, assuming you're using a YAML file for your dataset configuration:

```yaml
train: ../datasets/playing_cards/train/
val: ../datasets/playing_cards/val/

# Adding new style images into the existing classes
nc: 52
names: ['2C', '3C', ..., 'AD', 'Different_7']
```

Keep monitoring how your model performs with these adjustments. Fine-tuning the model with a mix of old and new styles may also help if you decide to introduce a new class specifically for the different 7s. Remember, the aim is to enhance the model's ability to generalize across variations within the same "logical" class in real-world conditions. Good luck! 🚀

**Reply by fluffytid** [reply]
Thanks Paula, another issue that I think has no solution but just in case.

When there are perfectly seen clipping cards I sometimes include them on my dataset, hoever on cases like these:

![image](https://github.com/ultralytics/ultralytics/assets/35581011/d0b102fc-9cfc-427e-be7b-d9012423a7f0)

Only the top cards of each pile could be really identified by the model.

The problem is, the percentage of card seen below another is a continuous scale, and there will be edge cases:

![image](https://github.com/ultralytics/ultralytics/assets/35581011/6fbbc6ec-7001-4685-ae47-0d854dadad66)

So  I am scared that if I include an image like that, if I tag a card that cannot be really identified I will create a false positive, but if I don't tag it, I will create false negatives. All in all I am not including any image with clipping cards unless really clearly seen. But on doing so, I am neglecting the model of data around a very common context case, but I see no way around it.

**Reply by pderrenger** [reply]
Hi there! 👋 I totally understand the challenge you're facing with partially visible cards. This is indeed a common dilemma in object detection tasks, especially with continuous scales of visibility.

In these situations, a practical approach could be to establish a **threshold of visibility** for labeling. For example, you might decide to label a card only if at least 50% of it is visible. This specific threshold can vary depending on what you observe works best for your model through experimentation. 

To further refine your model's ability to deal with these edge cases, you could consider **data augmentation techniques** that simulate card clippings, thus artificially increasing the diversity of your training dataset regarding partially visible objects. Here's a quick snippet to augment your images, considering you use YOLOv8's data augmentation capabilities:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Assuming you're using a custom dataset
model.train(
    data='path/to/your/dataset.yaml',
    epochs=50,
    imgsz=640,
    augment=True,  # Enable built-in data augmentation
)
```

Finally, it’s okay to occasionally have images that challenge the model—it helps it learn the limits of its detection capability. Your strategy of excluding unclear clippings unless they're really visible seems sensible. Balancing your dataset with a mix of clear and partially clipped cards, guided by a visibility threshold, could offer a way forward without leaning too much into false positives or negatives.

Hope this helps, and good luck with your model training! 🚀

**Comment by fluffytid** [comment]
Thank you.

I had no clue there was a data augmentation already built in, I have been
doing it myself with a program I develop and rotates the images. Where can
I find more information about what it exactly does?

**Reply by pderrenger** [reply]
Hi there! 😀

You can find detailed information about built-in data augmentation in YOLOv8 in our [documentation here](https://docs.ultralytics.com/tutorials/data-augmentation/). YOLOv8's data augmentation automatically applies various transformations like rotations, scaling, and flipping to your images during training, helping improve your model's robustness and ability to generalize across different scenarios.

If you're already performing custom augmentations, incorporating YOLOv8's built-in capabilities could further enhance your model's performance by introducing even more variation to your dataset. 

Here's a brief example on how to enable augmentation during training:
```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Start training with built-in augmentation enabled
model.train(data='path/to/your/dataset.yaml', epochs=50, imgsz=640, augment=True)
```

Combining YOLOv8's augmentation with your own can provide a comprehensive augmentation strategy without much hassle. Feel free to explore and experiment to see what works best for your specific use case. 

Happy modeling! 🚀

**Comment by fluffytid** [comment]
Thank you I will try to test it today, can you check the link to
documentation? it seems broken to me

**Reply by glenn-jocher** [reply]
Hello!

Thanks for pointing that out. It does seem that there was an issue with the link. Here's the correct one for our data augmentation documentation: [Data Augmentation in YOLOv8](https://docs.ultralytics.com/tutorials/data-augmentation/).

Please give it a try, and let us know how it goes. If you have any more questions or issues, feel free to ask!

Happy modeling! 🚀

**Comment by fluffytid** [comment]
When I run your code got some error regarding a thread starting too early.

I tried to run the code I usually do which is similar but from os with
augment=True in the end and did the first epoch. However in the end threw
this error:

C:\Users\fluff\PycharmProjects\pythonProject\venv39\Scripts\python.exe
C:\Users\fluff\PycharmProjects\pythonProject\cardetector\burner.py
New https://pypi.org/project/ultralytics/8.2.10 available  Update with 'pip
install -U ultralytics'
Ultralytics YOLOv8.0.112  Python-3.9.8 torch-2.0.1+cu117 CUDA:0 (NVIDIA
GeForce RTX 3080, 10240MiB)
yolo\engine\trainer: task=detect, mode=train, model=fullcardv7.pt,
data=../mergesymm/data.yaml, epochs=50, patience=50, batch=16, imgsz=640,
save=True, save_period=-1, cache=False, device=None, workers=8,
project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD,
verbose=True, seed=0, deterministic=True, single_cls=False, rect=False,
cos_lr=False, close_mosaic=0, resume=False, amp=True, fraction=1.0,
profile=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True,
split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7,
max_det=300, half=False, dnn=False, plots=True, source=None, show=False,
save_txt=False, save_conf=False, save_crop=False, show_labels=True,
show_conf=True, vid_stride=1, line_width=None, visualize=False,
augment=True, agnostic_nms=False, classes=None, retina_masks=False,
boxes=True, format=torchscript, keras=False, optimize=False, int8=False,
dynamic=False, simplify=False, opset=None, workspace=4, nms=False,
lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0,
warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5,
pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7,
hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0,
perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0,
copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml,
save_dir=runs\detect\train27

....
...

...

Logging results to runs\detect\train27
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances
Size
       1/50      9.55G      0.275     0.2158     0.8192         29
 640: 100%|██████████| 1004/1004 [06:28<00:00,  2.58it/s]
                 Class     Images  Instances      Box(P          R
 mAP50  mAP50-95):   0%|          | 0/77 [00:00<?, ?it/s]
Traceback (most recent call last):
  File
"C:\Users\fluff\AppData\Local\Programs\Python\Python39\lib\runpy.py", line
197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File
"C:\Users\fluff\AppData\Local\Programs\Python\Python39\lib\runpy.py", line
87, in _run_code
    exec(code, run_globals)
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\Scripts\yolo.exe\__main__.py",
line 7, in <module>
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\cfg\__init__.py",
line 398, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\engine\model.py",
line 371, in train
    self.trainer.train()
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\engine\trainer.py",
line 192, in train
    self._do_train(world_size)
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\engine\trainer.py",
line 368, in _do_train
    self.metrics, self.fitness = self.validate()
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\engine\trainer.py",
line 474, in validate
    metrics = self.validator(self)
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\torch\utils\_contextlib.py",
line 115, in decorate_context
    return func(*args, **kwargs)
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\engine\validator.py",
line 165, in __call__
    self.loss += model.loss(batch, preds)[1]
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\nn\tasks.py",
line 213, in loss
    return self.criterion(self.predict(batch['img']) if preds is None else
preds, batch)
  File
"C:\Users\fluff\PycharmProjects\pythonProject\venv39\lib\site-packages\ultralytics\yolo\utils\loss.py",
line 135, in __call__
    pred_distri, pred_scores = torch.cat([xi.view(feats[0].shape[0], self.no,
-1) for xi in feats], 2).split(
TypeError: 'NoneType' object is not iterable

**Reply by glenn-jocher** [reply]
Hi there! It looks like your model ran into a `TypeError` during the validation step after completing an epoch. This could be due to several reasons, including mismatches in the expected input dimensions or types within the model's loss function.

One quick check you can make is to ensure that all your data inputs are correctly formatted and that no `NoneType` data is being passed to the model inadvertently. It's also a good idea to ensure all the required dependencies and packages are up to date.

Regarding the validation step where the error is raised, examining the custom transformations or data manipulations applied might help. If the issue persists, simplifying the test case to pinpoint the cause by minimizing the custom settings could be beneficial.

Here's a snippet on how you might check if any `NoneType` input is passed during training, which might cause such issues:

```python
for batch in dataloader:
    if batch is None or any(x is None for x in batch):
        print("Found NoneType in batch data")
```

Please ensure the latest version is installed as per the update recommendation, and check if that resolves the error. Sometimes, such issues can be version-specific. If you continue having issues, feel free to share more details or the part of your code where you prepare your inputs!

**Reply by fluffytid** [reply]
Note that I run exactly the same images on my latest run but without augment=True and it worked flawlessly

**Reply by glenn-jocher** [reply]
Hello! It sounds like the issue might be specifically related to the `augment=True` setting. Augmentation can sometimes introduce data transformations that the model isn't handling as expected, especially if the augmentations affect image quality or important features in the images.

You could try adjusting the parameters of your augmentations to see if there's a particular transformation causing the issue, or simplify the augmentations to test incrementally. Here's an example of how you might modify the augmentation settings for clarity:

```python
model.train(data='path/to/your/dataset.yaml', epochs=50, imgsz=640, augment=True, degrees=0, translate=0.1, scale=0.5, shear=0)
```
This keeps some augmentations but reduces complexity. If this runs without errors, gradually reintroduce other parameters until you identify the problematic augmentation.

Hope this helps! Let me know how it goes. 😊

**Comment by fluffytid** [comment]
You gave me the same link?

**Reply by pderrenger** [reply]
Hello,

Apologies for the confusion. Here's the correct link for our data augmentation documentation: [Data Augmentation in YOLOv8](https://docs.ultralytics.com/tutorials/data-augmentation/).

Let us know if this works for you or if there's anything else we can help with!

Happy modeling! 🚀

**Comment by fluffytid** [comment]
had no idea you were an artificial intelligence until now, congratulations
to your developer!

**Reply by glenn-jocher** [reply]
Hello, thank you so much for your kind words! 🙏

I'm here to assist you further if you have any more questions about YOLOv8 or any other Ultralytics products. 

Have a great day and happy modeling! 🚀

**Reply by fluffytid** [reply]
Hi Glenn, if you have some info on the automated data augmentation that Paula has wrong on her database I'd like to read about it.

**Reply by glenn-jocher** [reply]
@fluffytid hello! You can find all the details about automated data augmentation for YOLOv8 in our comprehensive guide [here](https://docs.ultralytics.com/tutorials/data-augmentation/). This should give you a good overview of how it works and how you can utilize it for your projects!

If you need further assistance or have any questions, feel free to ask. Happy to help! 😊

**Comment by fluffytid** [comment]
Hi again Paula

I have noticed some discrepancy on my model when it is run through a stream
and when it is run through single images, and I believe it is because the
"stream=True" option, can you explain what that option does to me?

**Reply by glenn-jocher** [reply]
Hi there!

The `stream=True` option in YOLOv8 influences how the model processes the frames from a video or a live feed. When `stream=True` is set, the model reads one frame at a time and processes it instantly. This is efficient for real-time applications since it reduces memory usage by not holding all frames in memory at once. This can, however, potentially lead to discrepancies when compared to processing batched single images where each image is loaded and processed individually with possibly more resources available per image.

If you're seeing performance differences, it could indeed be due to the real-time constraints and the way frames are handled in streaming mode. You can adjust the frame processing configurations or test with `stream=False` to see if discrepancies reduce when processing videos as a series of batched images.

Hope that clarifies your query! Happy coding! 🚀
